pkg/controllers/virtualmachineinstance_controller.go

```diff
@@ -20,32 +20,44 @@ import (
 	"context"
 
 	"github.com/go-logr/logr"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 
 	v1 "kubevirt.io/api/core/v1"
+
+	"github.com/AlonaKaplan/kubesecondarydns/pkg/zonemgr"
 )
 
 // VirtualMachineInstanceReconciler reconciles a VirtualMachineInstance object
 type VirtualMachineInstanceReconciler struct {
 	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
-	// TODO ZoneManager zoneManager
+	Log         logr.Logger
+	Scheme      *runtime.Scheme
+	ZoneManager *zonemgr.ZoneManager
 }
 
 func (r *VirtualMachineInstanceReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
 	vmi := &v1.VirtualMachineInstance{}
 	err := r.Client.Get(context.TODO(), request.NamespacedName, vmi)
 	if err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("VMI is not found, updating the zone manager the VMI no longer exists")
+			err = r.ZoneManager.UpdateZone(request.NamespacedName, nil)
+			return ctrl.Result{}, err
+		}
+		r.Log.Error(err, "Error retrieving VMI")
+		// Error reading the object - requeue the request.
```

Comment:
we know it can happen ?

Reply:
`r.Client.Get(context.TODO(), request.NamespacedName, vmi)` has `err` as its return value, so I have to assume it may happen.

Reply:
this what bothers me, wouldn't it might re-enqueue forever in some specific error types
anyhow we can either live and see, or not re-enqueue and add specific cases upon need

Reply:
This is the pattern of the reconcile loop. It was generated by operator-sdk. https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/#reconcile-loop